Sortmerna testing for .gz input files

smruthi98 opened the pull request:
Testing code in Travis

smruthi98 committed:
Create DB

smruthi98 committed:
1. README.md - Included documentation for SortMeRNA
2. Sortmerna.py - Filtering ribosomal reads from metaranscriptomic samples.Generates 2 artifacts:ribsomal and non-ribsomal reads for each sample passed
3. test_sortmerna.py - Unittest for sortmerna.py.Current SortMeRNA version(v2.1) does not accept compressed (.gz) files. Databases and index  are stored under sortmerna/databases

Databases and index files for rRNA for unittest
Adding QC_SORTMERNA_DB_DP as environ variable
Included cython installation in travis.yml

smruthi98 committed:
Fixed minor linting issues

smruthi98 committed:
Update qp_shogun/sortmerna/README.md

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/README.md

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/sortmerna.py

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/utils.py

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/utils.py

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/utils.py

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/sortmerna.py

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/sortmerna.py

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Compressed DB, updated documentation and made the suggested changes

Updated travis.yml to include cython and uncompressing db files
Compressed databases to .tar.bz2 format
Removed .DS_Store files
Included params

smruthi98 committed:
Merge branch 'sortmerna' of https://github.com/smruthi98/qp-shogun into sortmerna

smruthi98 committed:
Update .gitignore

smruthi98 committed:
Delete .DS_Store

smruthi98 committed:
Delete .DS_Store

smruthi98 committed:
Delete .DS_Store

smruthi98 committed:
Delete .DS_Store

smruthi98 committed:
updated cython in travis.yml

smruthi98 committed:
Merge branch 'sortmerna' of https://github.com/smruthi98/qp-shogun into sortmerna

smruthi98 committed:
Update .travis.yml

smruthi98 committed:
Update .gitignore

smruthi98 committed:
Update test_sortmerna.py

smruthi98 committed:
Added conda-forge channel in line 16

build failed at libgfortran earlier

smruthi98 committed:
Compressed index files and removed extras

smruthi98 committed:
Update .travis.yml

smruthi98 committed:
updatd params

smruthi98 committed:
fixed sytanx error

smruthi98 committed:
fixed linting issues

smruthi98 committed:
Update __init__.py

smruthi98 committed:
modified format_params

smruthi98 committed:
updating param order to match sorted list

smruthi98 committed:
removed additional space in params and removed unused funcs in sortmerna/utils.py

smruthi98 committed:
Update test_sortmerna.py

smruthi98 committed:
Updating to latest databases

smruthi98 committed:
DB update for new version

Updated databses to reflect the most recent, can also be used in case we decide to switch to the newer SMR version

smruthi98 committed:
modifying utils uncompress .gz files

smruthi98 committed:
Merge remote-tracking branch 'upstream/master' into sortmerna

smruthi98 committed:
Update utils.py

smruthi98 committed:
Updated utils.py to compress SMR output file

ElDeveloper commented on qp_shogun/sortmerna/utils.py:
```diff
@@ -27,13 +60,17 @@ def _per_sample_ainfo(
             if exists(fname):
                 if fname.endswith('R1.fastq'):
                     ftype = 'raw_forward_seqs'
+                    # zipping output files
+                    fname_compressed = _gzip_compress(fname)
                 elif fname.endswith('R2.fastq'):
                     ftype = 'raw_reverse_seqs'
+                    # zipping output files
+                    fname_compressed = _gzip_compress(fname)
                 else:
                     # this should never happen and it's not really possible
                     # to reproduce so no tests!
                     raise ValueError('File %s has an unexpected name' % fname)
-                files.append((fname, ftype))
+                files.append((fname_compressed, ftype))
```
Since both branches imply compression, this can be rewritten like this:

```python
                  if fname.endswith('R1.fastq'):
                      ftype = 'raw_forward_seqs'
                  elif fname.endswith('R2.fastq'):
                      ftype = 'raw_reverse_seqs'
                  else:
                      # this should never happen and it's not really possible
                      # to reproduce so no tests!
                      raise ValueError('File %s has an unexpected name' % fname)

                  # zipping output files
                  fname_compressed = _gzip_compress(fname)
                  files.append((fname_compressed, ftype))
```

ElDeveloper commented on qp_shogun/sortmerna/utils.py:
```diff
@@ -10,11 +10,44 @@
 from qiita_client import ArtifactInfo
 
 
+# Returns a file list
+def _gzip_uncompress(input_filelist):
+    input_filelist_uncompressed = []
+    for input_file in input_filelist:
+        if input_file.endswith('.fastq.gz'):
+            filename_uncompressed = input_file.replace(
+                '.fastq.gz', '.fastq')
+            # Uncompressing and reading the .gz file
+            input = gzip.GzipFile(input_file, 'rb')
+            s = input.read()
+            input.close()
+            # Writing uncompressed data into a file
+            output = open(filename_uncompressed, 'wb')
+            output.write(s)
+            output.close()
+            # Return the .fastq filename
+            input_filelist_uncompressed.append(filename_uncompressed)
+        else:
+            raise ValueError('File %s has an unexpected name' % input_file)
+    return input_filelist_uncompressed
+
+
+def _gzip_compress(input_filename):
+    f_in = open(input_filename, 'rb')
+    output_filename = input_filename + '.gz'
+    f_out = gzip.open(output_filename, 'wb')
+    f_out.writelines(f_in)
+    f_out.close()
+    f_in.close()
+    return output_filename
+
```
This looks fine, except it might be better to do a command-call to [pigz](https://zlib.net/pigz/) for parallel compression/decompression. It could use the same number of threads as would be used for sortmerna. Any thoughts? Otherwise this could be very slow for large files.

If we don't want to use pigz, then we might want to read and write data in chunks i.e. passing the number of bytes to the read method. I don't think it would be good to read the whole file into memory.

smruthi98 replied:
Modified code to call pigz for compression/decompression within the genrate_smr command

ElDeveloper reviewed (changes requested):
Just two comments, looks good otherwise. Thanks @smruthi98!

smruthi98 committed:
Using pigz within SMR to uncompress and compress files

Removed smr utils file

smruthi98 committed:
Added -m parameter

Modified the default memory setting from 1G to 32G to potentially avoid splitting of files. Updated README doc as well to reflect it.

smruthi98 committed:
Update test_sortmerna.py

ElDeveloper commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -79,29 +78,73 @@ def generate_sortmerna_commands(forward_seqs, reverse_seqs, map_file,
+    # compressed files currently
 
     for run_prefix, sample, f_fp, r_fp in samples:
-        cmds.append('sortmerna --ref %s --reads %s '
```
For the command executions using `&&` instead of semicolons is preferable. This will help us identify errors on the commands that fail. Otherwise if `unpigz` failed we would probably see an error from `sortmerna` saying that the input file was not found or something along those lines.

Second, since the command is pretty much the same for forward and reverse files, I suggest setting that as a template and doing the formatting via a dictionary. For example like this:

```python
template = ("unpigz -p {threads -c {ip} > {ip_unpigz} && "
            "sortmerna --ref {ref_db} --reads {ip_unpigz} ...")
arguments = {'threads': threads, 'ref_db': RNA_REF_DB}

for run_prefix, sample, f_fp, r_fp in samples:
    prefix_path = join(out_dir, run_prefix)

    for index, fp in enumerate([f_fp, r_fp]):
        # if the reverse filepath is not present ignore it
        if fp is None:
            continue

        arguments['ip'] = fp
        arguments['ip_unpigz'] = fp.replace('.fastq.gz', '.fastq')
        arguments['smr_r_op'] = prefix_path + '.ribosomal.R%d' % (index + 1)
        arguments['smr_nr_op'] = prefix_path + '.nonribosomal.R%d' % (index + 1)
        arguments['smr_r_op_gz'] = arguments['smr_r_op'] + '.gz'
        arguments['smr_nr_op_gz'] = arguments['smr_nr_op'] + '.gz'

        cmds.append(template.format(**arguments))
```

smruthi98 replied:
Ah, I guess I saw this after I pushed the modified code to update memory requirements.

ElDeveloper replied:
Ping on this comment.

smruthi98 replied:
Done!

ElDeveloper reviewed (changes requested):
Thanks so much for using pigz, that's great! 🐷  Do we need to add this to the list of dependencies to ensure the program is installed before we can run the plugin?

smruthi98 committed:
Update test_sortmerna.py

smruthi98 committed:
Update sortmerna.py

smruthi98 committed:
Update sortmerna.py

smruthi98 committed:
Update sortmerna.py

smruthi98 committed:
Update test_sortmerna.py

smruthi98 committed:
Update sortmerna.py

smruthi98 committed:
Testing if SMR automatically appends file extension

smruthi98 committed:
Yep SMR adds file extensions

smruthi98 committed:
Update sortmerna.py

smruthi98 committed:
SMR v2.1b does not accept  > 8192M

smruthi98 committed:
Adjusting memory for db+input

smruthi98 committed:
Update test_sortmerna.py

smruthi98 committed:
Update test_sortmerna.py

smruthi98 committed:
Updating to inlcude && instad of ; in unpigz

smruthi98 committed:
Incoorporating && in generate SMR command

smruthi98 committed:
Memory requirements had to be modified

smruthi98 committed:
Max memory allowed 7976 M

ElDeveloper reviewed (approved):
Just one comment left that I think should be addressed.

smruthi98 committed:
updated SMR command as suggested

smruthi98 committed:
Update sortmerna.py

smruthi98 committed:
server error pushing it again

ElDeveloper reviewed (approved):
Thanks so much @smruthi98 🎉 :shipit:
